### PR TITLE
Darray free

### PR DESCRIPTION
**Pull request by ubitux**

### CODE REVIEW

File: libnopegl/src/hmap.c [hunk 1]
```diff
@@ -38,13 +38,13 @@ struct hmap {
     size_t size;
     size_t mask;
     size_t count; // total number of entries
-    user_free_func_type user_free_func;
+    hmap_user_free_func_type user_free_func;
```
Comment (mbouron) [inline libnopegl/src/hmap.c:41]: Can we define `ngli_user_free_func_type` (or even ngli_user_freep_func_type if that makes sense , that would in line with our current `ngli_freep_func_type`) in `utils.h` and use it for both hmap and darray ?

File: libnopegl/src/darray.c [hunk 2]
```diff
@@ -128,13 +137,25 @@ void *ngli_darray_get(const struct darray *darray, size_t index)
     return darray->data + index * darray->element_size;
 }
 
+static void invalidate_range(struct darray *darray, size_t index, size_t count)
+{
+    if (!darray->user_free_func)
+        return;
+    const size_t end = index + count;
```
Comment (mbouron) [inline libnopegl/src/darray.c:144]: You ll probably want the same check as it `remove_range`: `ngli_assert(end <= darray->count);`

File: libnopegl/src/backends/vk/bindgroup_vk.c [hunk 3]
```diff
@@ -241,6 +258,9 @@ int ngli_bindgroup_vk_init(struct bindgroup *s, const struct bindgroup_params *p
     ngli_darray_init(&s_priv->texture_bindings, sizeof(struct texture_binding_vk), 0);
     ngli_darray_init(&s_priv->buffer_bindings, sizeof(struct buffer_binding_vk), 0);
 
+    ngli_darray_set_free_func(&s_priv->texture_bindings, unref_texture_binding, NULL);
+    ngli_darray_set_free_func(&s_priv->buffer_bindings, unref_buffer_binding, NULL);
```
Comment (mbouron) [inline libnopegl/src/backends/vk/bindgroup_vk.c:262]: nit: personal preference, i would pack the init + set_free_func per array.

File: libnopegl/src/hud.c [hunk 4]
```diff
@@ -952,9 +952,21 @@ static int create_widget(struct hud *s, enum widget_type type, const void *user_
     return 0;
 }
 
+static void delete_widget(void *user_arg, void *data)
```
Comment (mbouron) [inline libnopegl/src/hud.c:955]: `free_widget` ?

**Review by mbouron — APPROVED** [review]
Except from my comments, LGTM.

File: libnopegl/src/hud.c (hunk 4 above)
Comment (ubitux) [reply libnopegl/src/hud.c:955]: it's doing both an uninit and then a free so I wanted a common term 🤷

Comment (mbouron) [reply libnopegl/src/hud.c:955]: `reset` then ? we don't `delete` in the rest of the code base

File: libnopegl/src/hmap.c (hunk 1 above)
Comment (ubitux) [reply libnopegl/src/hmap.c:41]: done

File: libnopegl/src/darray.c (hunk 2 above)
Comment (ubitux) [reply libnopegl/src/darray.c:144]: added

File: libnopegl/src/hud.c (hunk 4 above)
Comment (ubitux) [reply libnopegl/src/hud.c:955]: do you want me to rename all the other darray callbacks as well?

Comment (mbouron) [reply libnopegl/src/hud.c:955]: Nope. AFAIK, delete_ is only used here.

Comment (ubitux) [reply libnopegl/src/hud.c]: done

**Pull request merged**